packages/require/lib/compiler.ts

```diff
@@ -0,0 +1,94 @@
+import path from "path";
+import TruffleError from "@truffle/error";
+
+import type { CreateOptions, Service } from "ts-node";
+import TruffleConfig from "@truffle/config";
+import { readFileSync, existsSync } from "fs";
+
+import originalRequire from "original-require";
+
+import vm from "vm";
+
+/**
+ * @returns compiler function from user's installation of `ts-node` if script is
+ * a ts or tsx file, otherwise `undefined`.
+ */
+export function compile(conf: TruffleConfig, scriptPath: string): string {
+  const source = readFileSync(scriptPath, { encoding: "utf-8" });
+
+  // we only compile TS files, so just return the source unless we are dealing
+  // with one of those
+  if (![".ts", ".tsx", ".cts"].includes(path.extname(scriptPath))) {
+    return source;
+  }
+
+  const compilationService = _getOrCreateCompilationService(conf, scriptPath);
+
+  return compilationService.compile(source, scriptPath);
+}
+
+export function registerTsNode(
+  conf: TruffleConfig,
+  scriptPath: string,
+  context: Object
+) {
+  // we only compile TS files, so bail unless we are dealing with one of those;
+  if (![".ts", ".tsx", ".cts"].includes(path.extname(scriptPath))) {
+    return;
+  }
+
+  const code = `
+    const tsNode = require("ts-node");
+    const registrationOptions = {
+      cwd: ${JSON.stringify(conf.working_directory.toString())},
+      compilerOptions: {
+        inlineSourceMap: true
+      }
+    };
+    tsNode.register(registrationOptions);
+    `;
+  const script = new vm.Script(code);
+  script.runInContext(context);
+}
+
+let _compilationService: Service | null = null;
+
+function _getOrCreateCompilationService(
+  conf: TruffleConfig,
+  scriptPath: string
+): Service {
+  if (!_compilationService) {
+    try {
+      const createOptions: CreateOptions = {
+        cwd: path.dirname(scriptPath),
+        esm: false,
+        compilerOptions: {
+          inlineSourceMap: true
+        }
+      };
+
+      _compilationService = originalRequire("ts-node").create(createOptions);
+
+      const globalDeclarationsPath = existsSync(
+        path.join(__dirname, "scriptGlobals.ts")
+      )
+        ? path.join(__dirname, "scriptGlobals.ts")
+        : path.join(__dirname, "..", "scriptGlobals.ts");
```

Comment:
This ternary was necessary to make it so I could run truffle via `cli.js` as well as via the bundle.

Reply:
> I wonder if our webpack process should tweaked to either put the compiled source in the "correct" relative location, or update these lines during bundling to point to the right place.

I think the detail that you might be missing here is that I don't want the compiled source. I need to be able to read the _uncompiled_ source and pass it through the compiler host to augment the existing global types.

I started looking at the road of a `raw` loader for this file so that I could read in just the file contents, but then I'd still need to handle the unbundled case separately. Otherwise I think this is a bit more self-contained of a solution as-is than adding a "magic" var (aka a hidden input) that's set by the webpack bundler process.

As for the idiom of trying to read the file and handling the exception, I suppose I can move this into a separate `_getGlobalContextSource()` function and then follow that idiom there. I was just a bit lazy doing it inline, and I wanted a `const` rather than a `let` variable declaration.

Reply:
> I need to be able to read the uncompiled source and pass it through the compiler host to augment the existing global types.

I'll add a comment to clarify this as well, as it's a subtle but important issue. My desire was to make use of a `.d.ts` file here rather than a `.ts` file, but there's no good way to hand a `.d.ts` file to the compiler host to modify the global context for all files going forward. You can point the compiler at it with a `typeRoots` value in the `compilerOptions` but then that will override the default `typeRoots`, or any `typeRoots` defined by the user's `tsconfig.json`.